File: enaible-ts/src/runtime/context.ts
/**
 * Workspace discovery utilities for the Enaible CLI.
 */

import fs from 'fs';
import path from 'path';
import os from 'os';

const SHARED_SENTINEL = path.join('shared', 'core', 'base', 'analyzer_registry.py');
const SENTINEL_RELATIVE = path.join('core', 'base', 'analyzer_registry.py');
const DEFAULT_SHARED_HOME = path.join(os.homedir(), '.enaible', 'workspace', 'shared');

/**
 * Resolved workspace paths required by the CLI.
 */
export interface WorkspaceContext {
  repoRoot: string;
  sharedRoot: string;
  artifactsRoot: string;
}

/**
 * Get path from environment variable if it exists.
 */
function envPath(...keys: string[]): string | null {
  for (const key of keys) {
    const value = process.env[key];
    if (value) {
      const candidate = path.resolve(value.replace(/^~/, os.homedir()));
      if (fs.existsSync(candidate)) {
        return candidate;
      }
    }
  }
  return null;
}

/**
 * Generate candidate roots from a starting path.
 */
function* candidateRoots(start: string): Generator<string> {
  yield start;
  let current = start;
  let parent = path.dirname(current);
  while (parent !== current) {
    yield parent;
    current = parent;
    parent = path.dirname(current);
  }
}

/**
 * Find the shared root directory.
 */
function findSharedRoot(): string | null {
  // Explicit env override
  const envShared = envPath('ENAIBLE_SHARED_ROOT');
  if (envShared && fs.existsSync(path.join(envShared, SENTINEL_RELATIVE))) {
    return envShared;
  }

  // Installed workspace copy (created by installer)
  if (fs.existsSync(path.join(DEFAULT_SHARED_HOME, SENTINEL_RELATIVE))) {
    return DEFAULT_SHARED_HOME;
  }

  return null;
}

/**
 * Find the repository root directory.
 */
function findRepoRoot(start: string | null = null, requireShared: boolean = true): string {
  const envRepo = envPath('ENAIBLE_REPO_ROOT');
  if (envRepo) {
    if (!requireShared || fs.existsSync(path.join(envRepo, SHARED_SENTINEL))) {
      return envRepo;
    }
  }

  const searchStart = path.resolve(start ?? process.cwd());
  if (!requireShared) {
    return searchStart;
  }

  for (const candidate of candidateRoots(searchStart)) {
    if (fs.existsSync(path.join(candidate, SHARED_SENTINEL))) {
      return candidate;
    }
  }

  throw new Error(
    'Unable to locate repository root; set ENAIBLE_REPO_ROOT or run inside a checkout.'
  );
}

/**
 * Resolve the artifacts root directory.
 */
function resolveArtifactsRoot(repoRoot: string): string {
  let artifacts = envPath('ENAIBLE_ARTIFACTS_DIR', 'ENAIBLE_ARTIFACTS_ROOT');
  if (!artifacts) {
    artifacts = path.join(repoRoot, '.enaible');
  }
  if (!fs.existsSync(artifacts)) {
    fs.mkdirSync(artifacts, { recursive: true });
  }
  return artifacts;
}

/**
 * Resolve workspace paths.
 */
export function loadWorkspace(start: string | null = null): WorkspaceContext {
  let sharedRoot = findSharedRoot();

  // If we have a packaged/shared copy, we can relax repo discovery to the current tree.
  const repoRoot = findRepoRoot(start, sharedRoot === null);

  if (!sharedRoot) {
    sharedRoot = path.join(repoRoot, 'shared');
    if (!fs.existsSync(sharedRoot)) {
      throw new Error(
        `Shared analyzers folder missing at ${sharedRoot}; repository may be incomplete.`
      );
    }
  }

  const artifactsRoot = resolveArtifactsRoot(repoRoot);

  return {
    repoRoot,
    sharedRoot,
    artifactsRoot,
  };
}

/**
 * Public helper to locate the shared workspace root.
 */
export { findSharedRoot };

File: ai-assisted-workflows/tools/enaible-rs/crates/enaible-core/src/context.rs
use std::env;
use std::path::{Path, PathBuf};
use thiserror::Error;

/// Workspace paths required by the CLI
#[derive(Debug, Clone)]
pub struct WorkspaceContext {
    pub repo_root: PathBuf,
    pub shared_root: PathBuf,
    pub artifacts_root: PathBuf,
}

#[derive(Error, Debug)]
pub enum WorkspaceError {
    #[error("Unable to locate repository root; set ENAIBLE_REPO_ROOT or run inside a checkout")]
    RepoRootNotFound,

    #[error("Shared analyzers folder missing at {0}; repository may be incomplete")]
    SharedRootMissing(PathBuf),

    #[error("IO error: {0}")]
    Io(#[from] std::io::Error),
}

const SHARED_SENTINEL: &str = "shared/core/base/analyzer_registry.py";
const SENTINEL_RELATIVE: &str = "core/base/analyzer_registry.py";

/// Get a path from environment variable
fn env_path(keys: &[&str]) -> Option<PathBuf> {
    for key in keys {
        if let Ok(value) = env::var(key) {
            let candidate = PathBuf::from(value)
                .canonicalize()
                .ok()?;
            if candidate.exists() {
                return Some(candidate);
            }
        }
    }
    None
}

/// Find the shared root directory
pub fn find_shared_root() -> Option<PathBuf> {
    // Explicit env override
    if let Some(env_shared) = env_path(&["ENAIBLE_SHARED_ROOT"]) {
        if env_shared.join(SENTINEL_RELATIVE).exists() {
            return Some(env_shared);
        }
    }

    // Installed workspace copy
    let default_shared_home = home::home_dir()
        .map(|h| h.join(".enaible").join("workspace").join("shared"))
        .unwrap_or_default();

    if default_shared_home.join(SENTINEL_RELATIVE).exists() {
        return Some(default_shared_home);
    }

    // Will fallback to repo-relative detection later
    None
}

/// Find the repository root
fn find_repo_root(start: Option<&Path>, require_shared: bool) -> Result<PathBuf, WorkspaceError> {
    // Check environment variable first
    if let Some(env_repo) = env_path(&["ENAIBLE_REPO_ROOT"]) {
        if !require_shared || env_repo.join(SHARED_SENTINEL).exists() {
            return Ok(env_repo);
        }
    }

    // Start from current directory or provided path
    let search_start = match start {
        Some(p) => p.to_path_buf(),
        None => env::current_dir()?,
    }.canonicalize()?;

    if !require_shared {
        return Ok(search_start);
    }

    // Walk up directory tree looking for sentinel file
    let mut current = search_start.as_path();
    loop {
        if current.join(SHARED_SENTINEL).exists() {
            return Ok(current.to_path_buf());
        }

        match current.parent() {
            Some(parent) => current = parent,
            None => break,
        }
    }

    // Also check from the binary location
    if let Ok(exe_path) = env::current_exe() {
        let mut current = exe_path.as_path();
        loop {
            if current.join(SHARED_SENTINEL).exists() {
                return Ok(current.to_path_buf());
            }

            match current.parent() {
                Some(parent) => current = parent,
                None => break,
            }
        }
    }

    Err(WorkspaceError::RepoRootNotFound)
}

/// Resolve artifacts root directory
fn resolve_artifacts_root(repo_root: &Path) -> PathBuf {
    env_path(&["ENAIBLE_ARTIFACTS_DIR", "ENAIBLE_ARTIFACTS_ROOT"])
        .unwrap_or_else(|| repo_root.join(".enaible"))
}

/// Load workspace context from the current environment
pub fn load_workspace(start: Option<&Path>) -> Result<WorkspaceContext, WorkspaceError> {
    let shared_root = find_shared_root();

    // If we have a packaged/shared copy, we can relax repo discovery
    let repo_root = find_repo_root(start, shared_root.is_none())?;

    let shared_root = match shared_root {
        Some(root) => root,
        None => {
            let root = repo_root.join("shared");
            if !root.exists() {
                return Err(WorkspaceError::SharedRootMissing(root));
            }
            root
        }
    };

    let artifacts_root = resolve_artifacts_root(&repo_root);

    // Create artifacts directory if it doesn't exist
    std::fs::create_dir_all(&artifacts_root).ok();

    Ok(WorkspaceContext {
        repo_root,
        shared_root,
        artifacts_root,
    })
}